test/dynamo/test_repros.py

```diff
@@ -1959,6 +1959,77 @@ def f(x, y):
+        #   However, sometimes it is possible for those GraphArgs to be
+        #   utilized in shape guards (you could imagine this happening if
+        #   dynamo poked some shape variables without recording them in the
+        #   graph.)  If we prune those GraphArgs, we get a
+        #   "s1 not in ..." error as we can no longer codegen the
+        #   requested guards.
+        #
+        # - But in practice, Dynamo usually traces size accesses into the
+        #   graph, preventing the GraphArg from getting pruned.  So how
+        #   come we were running into this in practice with hf_Reformer?
+        #   The answer is checkpointing!
+        #
+        # This brings us to the following test case.  Here's what it does:
+        #
+        # 1. It traces some operations, and then checkpoints before inlining
+        #    the function call to g
+        #
+        # 2. g traces some more operations (triggering the shape guard
+        #    to be created), but then it graph breaks
+        #
+        # 3. Because you can't graph break in an inlining function, we roll
+        #    back to the outer checkpoint ("undoing" the operation that
+        #    induced the shape guard) and then immediately generate a
+        #    subgraph at that point.
+        #
+        # If we failed to checkpoint the ShapeEnv, it can still have guards
+        # from the aborted speculation, which we will then still attempt to
+        # codegen.
+        #
+        # There's an additional nuance: suppose x is used but y is not.
+        # If you create a guard like y == x * 2, you will accidentally avoid
+        # the "s1 not in ..." error, as y will get substituted with x * 2,
+        # but x is still a GraphArg (it's used) and you don't end up with
+        # the error.  This is why we must show y + y == x, not vice versa.
+        # Similarly, it is also why we must not do a simple guard like x == y
+        #
+        # Can we actually demonstrate that checkpointing the ShapeEnv is
+        # necessary?  It's not so easy to induce this case.  Dynamo is very
+        # eager about adding locals to GraphArgs; any local that is in scope,
+        # even if it isn't used, is added to GraphArgs (see also
+        # https://github.com/pytorch/torchdynamo/issues/1925 ).  So long
+        # as Dynamo eagerly guards in this way, we have an invariant that
+        # all locals are guaranteed to show up in GraphArgs before the
+        # inlining function call, in which case we will always have enough
+        # information to codegen our guards so long as we don't prune the
+        # unused GraphArgs away (and indeed, the direct fix for this bug
+        # was to make sure we use original GraphArgs).  Non locals,
+        # conversely, typically are static, and so won't have guards allocated
+        # for them.  That being said, there may still be a way to trigger
+        # this error.
```

Comment:
This deserves its own test to protect this invariant, we take it as a matter of course for now, but it changing subtly under the hood could cause really nasty to debug issues. Let's file an issue to test this (specifically, the exhaustive guarding of locals).

Reply:
This isn't a variant. I just failed to come up with a test case that could actually trigger the more naughty fix.

The proper fix is to checkpoint shape env properly. But I'm hoping we can put that off a bit.

Reply:
It turns out that you can trigger the edge case discussed here with unspecialized ints, and in fact quite easily.